docs/spec/connection-handshake/L1_2.md

```diff
@@ -0,0 +1,626 @@
+
+```golang
+type ConnectionEnd interface {
+  state                             ConnectionState
+  remoteParty                       ConnectionParty
+  localParty                        ConnectionParty
+  version                           []String
+}
+```
+
+
+4. A `ConnProof` and `ClientProof` are two data types that abstracts over the details of a cryptographic proof that any module can create, and another module can verify.
+Intuitively, a proof is helpful for some module to guarantee that it is in a certain state.
+We are concerned with connection proofs (`ConnProof` type) and client proofs (`ClientProof`) here.
+
+```golang
+type ConnProof struct {
+  proof CommitmentProof // The proof for the connection that the remote party claims to have.
+}
+```
+
+```golang
+type ClientProof struct {
+  proof CommitmentProof // The proof for the client that the remote party claims to have.
+  height uint64 // The height which the client (on remote party) claims having.
+}
+```
+
+We now restate the connection handshake problem in a slightly more precise (or lower level) formulation:
+
+__A connection handshake enables two ICS3 modules to agree on a consistent view of their chain state, to verify each other's state, and to allocate a new connection.__
+
+
+###### Guarantees
+We refine the safety guarantee **[ICS3-Seq-1-Agreement]** (defined in the [sequential problem statement](#sequential-problem-statement)) by splitting this into three complementary safety properties:
+
+- **[ICS3-Proto-1-ConnectionUniqueness]**
+If a module establishes a `ConnectionEnd` `e`, no handshake has previously finished for `e`.
+
+- **[ICS3-Proto-2-ConnectionIntegrity]**
+If any two modules establish a connection `e`, then either one module or the other or both modules accepted (i.e., initialized with) `e`.
+
+- **[ICS3-Proto-3-StateConsistency]**
+If any two modules establish a connection, then the client in the first module is consistent with the state of the second module.
+
+The liveness property below is equivalent to **[ICS3-Seq-2-Termination]** (the livenes property in the sequential specification above):
+
+- **[ICS3-Proto-5-Termination]**
+If the two parties involved in a connection handshake are correct, then the connection handshake eventually terminates.
+Specifically, termination implies that each module allocates in the local state a new `ConnectionEnd` object with `state` field set to `OPEN`.
+
+
+_Remarks_:
+
+- Uniqueness property essentially provides a safeguard against overwritting a connection in the store with some new set of parameters.
+
+- One of the mechanisms that enforces the connection integrity property consists of cryptographic proofs, specifically the `ConnProof` type, which asserts that a certain chain stores a certain connection in a certain state.
+
+- The `ClientProof` type helps guarantee the state consistency property, by asserting that the client of a certain chain is updated to a certain height of the remote chain.
```

Comment:
But as far as I'm aware there is no requirement in ICS3 on how __recent__ the consensus state on a client has to be. It would be useful to have a freshness guarantee, but is this enforced by `verifyClientConsensusState`, or where?